// @intuit/judo/dist/logger.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.logger = void 0;var _colors = _interopRequireDefault(require("colors"));function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

class Logger {
  info(...args) {
    console.log(_colors.default.cyan('[INFO]    ', ...args));
  }
  logOutput(...args) {
    console.log(_colors.default.white('[OUTPUT]  ', ...args));
  }
  logStdout(...args) {
    console.log(_colors.default.green('[STDOUT]  ', ...args));
  }
  logStderr(...args) {
    console.log(_colors.default.yellow('[STDERR]  ', ...args));
  }
  warn(...args) {
    console.log(_colors.default.magenta('[WARN]    ', ...args));
  }
  error(...args) {
    console.log(_colors.default.red('[ERROR]   ', ...args));
  }
  success(...args) {
    console.log(_colors.default.green('[SUCCESS] ', ...args));
  }
  summary(...args) {
    console.log(_colors.default.yellow('[SUMMARY] ', ...args));
  }
  lineBreak() {
    console.log('');
  }}


const logger = new Logger();exports.logger = logger;
